### PR TITLE
Allow providing the "work" volume(mount) as part of the RunnerDeployment

### PR DESCRIPTION
**Pull request by genisd**
This allows providing a different `work` Volume.

This should be a cloud agnostic way of allowing the operator to use (for example) NVME backed storage.

This is a working example where the workDir will use the provided volume, additionally here docker is placed on the same NVME.
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: runner-2
spec:
  template:
    spec:
      dockerdContainerResources: {}
      env:
      - name: POD_NAME
        valueFrom:
          fieldRef:
            fieldPath: metadata.name
      # this is to mount the docker in docker onto NVME disk
      dockerVolumeMounts:
      - mountPath: /var/lib/docker
        name: scratch
        subPathExpr: $(POD_NAME)-docker
      - mountPath: /runner/_work
        name: work
        subPathExpr: $(POD_NAME)-work
      volumeMounts:
      - mountPath: /runner/_work
        name: work
        subPathExpr: $(POD_NAME)-work
      dockerEnv:
      - name: POD_NAME
        valueFrom:
          fieldRef:
            fieldPath: metadata.name
      volumes:
      - hostPath:
          path: /mnt/disks/ssd0
        name: scratch
      - hostPath:
          path: /mnt/disks/ssd0
        name: work
      nodeSelector:
       cloud.google.com/gke-nodepool: runner-16-with-nvme
      ephemeral: false
      image: ""
      imagePullPolicy: Always
      labels:
        - runner-2
        - self-hosted
      organization: yourorganization

```

### CODE REVIEW

File: controllers/runner_controller.go [hunk 1]
```diff
@@ -1123,3 +1148,21 @@ func removeFinalizer(finalizers []string, finalizerName string) ([]string, bool)
 
 	return result, removed
 }
+
+func workVolumePresent(items []corev1.Volume) (bool, int) {
```
Comment (genisd) [inline controllers/runner_controller.go:1152]: I could make this fn type agnostic, that it accepts `items []corev1.Volume` and `items []corev1.VolumeMount` like the fn below

File: controllers/runner_controller.go [hunk 2]
```diff
@@ -671,10 +671,29 @@ func (r *RunnerReconciler) newPod(runner v1alpha1.Runner) (corev1.Pod, error) {
 	runnerSpec := runner.Spec
 
 	if len(runnerSpec.VolumeMounts) != 0 {
+		// if operater provides a work volume mount, use that
```
Comment (genisd) [inline controllers/runner_controller.go:674]: An alternative to stripping/removing the work VolumeMount here is simply adding it here if it's not provided by the user.
From reading the code, the functions where it's added should be like a "default" template, so I opted for this.

I tried skipping it in that default template, but the needed context is not always present.
The user provided volumes is not available, but easily refactored. Only issue I couldn't overcome was the "registration-only" codepath was somehow different.

**Comment by genisd** [comment]
@mumoshu could you take a look at this please?
It's running on our own environment.

I think this solves https://github.com/actions-runner-controller/actions-runner-controller/issues/1007 as well, since now a full volume can be specified.

Like:
```
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  name: runner-2
spec:
  template:
    spec:
      volumes:
      - name: work
        emptyDir:
          medium: Memory
```

should work.

**Comment by mumoshu** [comment]
@genisd Thanks for your patience and contribution! I'll review this in a day or two.

File: controllers/runner_controller.go (hunk 2 above)
Comment (genisd) [reply controllers/runner_controller.go:675]: Once https://github.com/actions-runner-controller/actions-runner-controller/pull/1028 is merged I can simplify this pr and make it cleaner/neater.

**Review by mumoshu — APPROVED** [review]
I didn't really mean to call two weeks "a day or two" 😆 
After careful reading, this change looks straightforward enough and safe to be included in 0.21.0.
Sorry for overriding the milestone @toast-gear but I believe this is worth a lot!

**Comment by mumoshu** [comment]
@genisd Thanks for your contribution! It would be great if you could add some documentation on this feature to our README, so that people suffering from issues like #1007 would easily find your awesome enhancement 😄

**Comment by genisd** [comment]
Thank you, I'll definitely create a PR for documentation, it's indeed something I know i had not done yet.

I'm going to especially check things like if data gets left behind and such things, put a proper disclaimer next to the examples.
It's not an issue for us, but probably worth mentioning.

**Comment by edspencer** [comment]
Thanks @genisd (and @mumoshu) - looking forward to abandoning my janky custom solution from https://github.com/actions-runner-controller/actions-runner-controller/issues/1007 in favor of your new magic when 0.21 comes out 🙌🏻

**Pull request merged**